linera-core/src/client.rs

```diff
@@ -298,32 +307,105 @@ impl From<Infallible> for ChainClientError {
     }
 }
 
+// We never want to pass the DashMap references over an `await` point, for fear of
+// deadlocks.  The following construct will cause a (relatively) helpful error if we do.
```

Comment:
Should we create a safer `DashMap` library in `linera_base` then?

Reply:
Maybe. There's not a _lot_ we can do, without resorting to `tokio::sync::Mutex`-like task actors. I think that maybe we could remove the behaviour in which a mut ref into anywhere in the map (as opposed to a mut ref to the map itself, e.g. when inserting or deleting) locks the whole map, but I don't fully understand why that behaviour is there in the first place, which makes me cautious. 

Given more time we could probably concoct a concurrent map that only _ever_ locks one key at a time, but that's a more serious engineering project. 

My plan for now is to avoid returning locks from the core client as much as possible, making them explicit where they are returned (hence the `ChainGuard` types here), and for the user-facing `linera-client` library avoid returning them at all, encapsulating any operations that use them into functions.